pynamodb/asyncio/batch_write.py

```diff
@@ -0,0 +1,138 @@
+    BATCH_WRITE_PAGE_LIMIT,
+    PUT,
+    DELETE,
+    PUT_REQUEST,
+    DELETE_REQUEST,
+    UNPROCESSED_ITEMS,
+)
+from pynamodb.exceptions import PutError
+
+_T = typing.TypeVar("_T")
+BatchOperation = typing.Dict[str, typing.Any]
+SerializedItem = typing.Dict[str, typing.Any]
+
+
+class AsyncBatchWrite(typing.Generic[_T], typing.AsyncContextManager[typing.Self]):
+    """
+    Async context manager for batch write operations in DynamoDB.
+
+    This class provides an interface for batched write operations (put/delete)
+    that are executed asynchronously when the context manager exits.
+    """
+
+    def __init__(self, model: typing.Type[_T], auto_commit: bool = True):
+        self.model = model
+        self.max_operations = BATCH_WRITE_PAGE_LIMIT
+        self.pending_operations: typing.List[BatchOperation] = []
+        self.failed_operations: typing.List[typing.Any] = []
+        self.auto_commit = auto_commit
+
+    async def save(self, put_item: _T) -> None:
+        """
+        Queue an item for batch insertion.
+
+        Args:
+            put_item: The model instance to be inserted
+
+        Raises:
+            ValueError: If the model uses versioning
+        """
+        if put_item._version_attribute_name is not None:
+            raise ValueError(
+                "batch_write does not support versioned models. Use a transaction instead"
+            )
+        if len(self.pending_operations) == self.max_operations:
+            if self.auto_commit:
+                await self.commit()
+        self.pending_operations.append({"action": PUT, "item": put_item})
+
+    async def delete(self, del_item: _T) -> None:
+        """
+        Queue an item for batch deletion.
+
+        Args:
+            del_item: The model instance to be deleted
+
+        Raises:
+            ValueError: If the model uses versioning
+        """
+        if del_item._version_attribute_name is not None:
+            raise ValueError(
+                "batch_write does not support versioned models. Use a transaction instead"
+            )
+        if len(self.pending_operations) == self.max_operations:
+            if self.auto_commit:
+                await self.commit()
+        self.pending_operations.append({"action": DELETE, "item": del_item})
+
+    async def __aenter__(self) -> "AsyncBatchWrite[_T]":
+        """Enter the async context manager."""
+        return self
+
+    async def __aexit__(
+        self,
+        exc_type: typing.Optional[typing.Type[BaseException]],
+        exc_val: typing.Optional[BaseException],
+        exc_tb: typing.Optional[typing.Any],
+    ) -> None:
+        """Exit the async context manager and commit pending operations."""
+        await self.commit()
+
+    async def commit(self) -> None:
+        """
+        Commit all pending batch write operations.
+
+        Raises:
+            PutError: If the maximum retry attempts are exceeded
+        """
+
+        async def _tasks() -> typing.AsyncIterator[
```

Comment:
This might work better as a [TaskGroup](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup)?

Reply:
I've tried, the problem with the TaskGroup is that I couldn't figure out a way to retrieve the results back. Apparently I can't (??)